File: lib/channels/mirai/protocol.ts
import { z } from "zod"

/**
 * mirai 桥接协议(v1)。建连方向可选；入站始终指插件发往 Prayer。
 *
 * 设计取舍:
 * - 入站帧全部经 zod 校验 —— 对端是独立进程/独立语言栈,不能假定它守约;
 *   非法帧只丢弃并记 warn,绝不允许崩掉服务端(会连带断掉其它 client)。
 * - 所有帧带 v,便于将来不兼容演进时并存两版。
 * - 图片走 base64 内联:插件侧已 queryUrl 下载,Prayer 侧不再出网。
 */
export const PROTOCOL_VERSION = 1

/** 单帧上限:图片 base64 会撑大体积,超限说明对端未按约束压缩 */
export const MAX_FRAME_BYTES = 8 * 1024 * 1024

const imageSchema = z.object({
  data: z.string().min(1),
  mediaType: z.string().min(1),
})

/** 群内角色:与 OneBot 对齐,供反思识别人工回复 */
const roleSchema = z.enum(["owner", "admin", "member"])

const chatSchema = z.object({
  id: z.string().min(1),
  name: z.string().default(""),
})

/** 插件的连接后首帧：声明身份与所辖会话，Prayer 据此建路由。 */
export const helloFrameSchema = z.object({
  v: z.literal(PROTOCOL_VERSION),
  type: z.literal("hello"),
  clientId: z.string().min(1),
  botId: z.number().int().positive(),
  chats: z.array(chatSchema).default([]),
})

export const messageFrameSchema = z.object({
  v: z.literal(PROTOCOL_VERSION),
  type: z.literal("message"),
  /** 必须稳定:Prayer 侧据此去重,重连重放不应导致重复回答 */
  messageId: z.string().min(1),
  chatId: z.string().min(1),
  userId: z.string().min(1),
  senderRole: roleSchema.optional(),
  text: z.string().default(""),
  /** 消息链是否含 At(bot);由插件判定后显式上报,gateway 直接采信 */
  botMentioned: z.boolean().default(false),
  quoted: z.string().optional(),
  forwarded: z.string().optional(),
  images: z.array(imageSchema).optional(),
  ts: z.number().int().nonnegative().optional(),
})

/** 回查应答(对应服务端 request);data 形状由 action 决定,此处不窄化 */
export const responseFrameSchema = z.object({
  v: z.literal(PROTOCOL_VERSION),
  type: z.literal("response"),
  echo: z.string().min(1),
  data: z.unknown().optional(),
  error: z.string().optional(),
})

export const pongFrameSchema = z.object({
  v: z.literal(PROTOCOL_VERSION),
  type: z.literal("pong"),
})

export const inboundFrameSchema = z.discriminatedUnion("type", [
  helloFrameSchema,
  messageFrameSchema,
  responseFrameSchema,
  pongFrameSchema,
])

export type HelloFrame = z.output<typeof helloFrameSchema>
export type MessageFrame = z.output<typeof messageFrameSchema>
export type ResponseFrame = z.output<typeof responseFrameSchema>
export type InboundFrame = z.output<typeof inboundFrameSchema>

/** 出站:发群消息;replyToId 存在则引用回复 */
export interface SendFrame {
  v: typeof PROTOCOL_VERSION
  type: "send"
  chatId: string
  text: string
  replyToId?: string
}

/** 出站:回查请求,客户端须以同 echo 的 response 应答 */
export interface RequestFrame {
  v: typeof PROTOCOL_VERSION
  type: "request"
  echo: string
  action: "listChats" | "listMembers"
  params?: Record<string, unknown>
}

export interface PingFrame {
  v: typeof PROTOCOL_VERSION
  type: "ping"
}

export type OutboundFrame = SendFrame | RequestFrame | PingFrame

export interface ParseResult {
  frame?: InboundFrame
  error?: string
}

/**
 * 解析入站帧。任何异常都收敛成 error 字符串,调用方只丢弃该帧。
 * 不抛异常:一个畸形帧不应影响连接与其它 client。
 */
export function parseInboundFrame(raw: string): ParseResult {
  const bytes = Buffer.byteLength(raw, "utf8")
  if (bytes > MAX_FRAME_BYTES) {
    return { error: `帧过大(${bytes} > ${MAX_FRAME_BYTES})` }
  }
  let json: unknown
  try {
    json = JSON.parse(raw)
  } catch {
    return { error: "非法 JSON" }
  }
  const parsed = inboundFrameSchema.safeParse(json)
  if (!parsed.success) {
    // 只取首个 issue:完整 issues 可能很长,日志无需刷屏
    const first = parsed.error.issues[0]
    const path = first?.path.join(".") || "(root)"
    return { error: `帧校验失败 ${path}: ${first?.message ?? "unknown"}` }
  }
  return { frame: parsed.data }
}

export function encodeFrame(frame: OutboundFrame): string {
  return JSON.stringify(frame)
}

File: mirai-plugin/src/main/kotlin/Protocol.kt
package org.prayer.bridge

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

/**
 * 桥接协议 v1，与 Prayer 侧 lib/channels/mirai/protocol.ts 的 zod schema 一一对应。
 * 改动任一侧都必须同步另一侧，否则帧会被对端校验拒绝。
 */
const val PROTOCOL_VERSION = 1
const val MAX_FRAME_BYTES = 8 * 1024 * 1024

/**
 * 帧编解码集中到一处，client / server / 插件三方共用同一 Json 实例与编码逻辑，
 * 避免各自 new Json 造成配置漂移（如 encodeDefaults 不一致导致漏字段）。
 */
@OptIn(kotlinx.serialization.ExperimentalSerializationApi::class)
object FrameCodec {
    val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
        explicitNulls = false
    }

    fun encode(frame: HelloFrame): String = json.encodeToString(HelloFrame.serializer(), frame)
    fun encode(frame: MessageFrame): String {
        var bounded = frame
        var payload = json.encodeToString(MessageFrame.serializer(), bounded)
        while (payload.toByteArray(Charsets.UTF_8).size > MAX_FRAME_BYTES && !bounded.images.isNullOrEmpty()) {
            bounded = bounded.copy(images = bounded.images!!.dropLast(1).takeIf { it.isNotEmpty() })
            payload = json.encodeToString(MessageFrame.serializer(), bounded)
        }
        require(payload.toByteArray(Charsets.UTF_8).size <= MAX_FRAME_BYTES) { "Message exceeds frame limit" }
        return payload
    }
    fun encode(frame: ResponseFrame): String = json.encodeToString(ResponseFrame.serializer(), frame)
    fun encode(frame: PongFrame): String = json.encodeToString(PongFrame.serializer(), frame)

    /** 解析入站帧；畸形帧交由调用方按上下文处理（记录/忽略/拒绝）。 */
    fun decodeInbound(raw: String): InboundFrame =
        json.decodeFromString(InboundFrame.serializer(), raw).also {
            require(it.v == PROTOCOL_VERSION) { "Unsupported protocol version" }
        }
}

@Serializable
data class ChatDto(
    val id: String,
    val name: String = "",
)

@Serializable
data class ImageDto(
    /** base64，不带 data: 前缀 */
    val data: String,
    val mediaType: String,
)

/** 出站：连接后首帧，声明身份与所辖会话 */
@Serializable
data class HelloFrame(
    val v: Int = PROTOCOL_VERSION,
    val type: String = "hello",
    val clientId: String,
    val botId: Long,
    val chats: List<ChatDto> = emptyList(),
)

/** 出站：群消息 */
@Serializable
data class MessageFrame(
    val v: Int = PROTOCOL_VERSION,
    val type: String = "message",
    /** 必须稳定：Prayer 侧据此去重，重连重放不应导致重复回答 */
    val messageId: String,
    val chatId: String,
    val userId: String,
    val senderRole: String? = null,
    val text: String = "",
    val botMentioned: Boolean = false,
    val quoted: String? = null,
    val forwarded: String? = null,
    val images: List<ImageDto>? = null,
    val ts: Long? = null,
)

/** 出站：回查应答 */
@Serializable
data class ResponseFrame(
    val v: Int = PROTOCOL_VERSION,
    val type: String = "response",
    val echo: String,
    val data: JsonElement? = null,
    val error: String? = null,
)

@Serializable
data class PongFrame(
    val v: Int = PROTOCOL_VERSION,
    val type: String = "pong",
)

/**
 * 入站帧的宽松形态：一次解析出所有可能字段，按 type 分派。
 * 不用 sealed class + 多态序列化，是为了避免 kotlinx 的 classDiscriminator
 * 与本协议 type 字段语义冲突，也让未知 type 能安全忽略而非抛异常。
 */
@Serializable
data class InboundFrame(
    val v: Int,
    val type: String,
    // send
    val chatId: String? = null,
    val text: String? = null,
    val replyToId: String? = null,
    // request
    val echo: String? = null,
    val action: String? = null,
    val params: Map<String, String>? = null,
)

/** 成员列表回查的返回项；字段名对齐 OneBot 习惯，便于后台复用展示 */
@Serializable
data class MemberDto(
    @SerialName("user_id") val userId: Long,
    @SerialName("nickname") val nickname: String,
    @SerialName("card") val card: String = "",
    @SerialName("role") val role: String = "member",
)
